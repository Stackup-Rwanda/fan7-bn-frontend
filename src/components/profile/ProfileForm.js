import React, { Component } from 'react';
import PropTypes from 'prop-types';
import moment from 'moment';
import './styles/profileForm.scss';

class ProfileForm extends Component {
  constructor(props) {
    super(props);
    this.initialState = {
      user: null,
    };
    this.state = this.initialState;
    this.handleSubmit = this.handleSubmit.bind(this);
    this.handleChange = this.handleChange.bind(this);
  }

  handleSubmit(event) {
    event.preventDefault();
    const { updateProfile } = this.props;
    const { user } = this.state;
    updateProfile(user);
  }

  handleChange(event) {
    const { name, value } = event.target;
    this.setState(prevState => ({
      user: {
        ...prevState.user,
        [name]: value,
      },
    }));
  }

  render() {
    const { profile, loading } = this.props;
    const {
      first_name: firstName,
      last_name: lastName,
      user_name: userName,
      gender,
      dob,
      phone,
      address,
      country,
      prefered_language: preferedLanguage,
      prefered_currency: preferdCurrency,
      company,
      department,
    } = profile;

    return (
      <form className="form" onSubmit={this.handleSubmit}>
        <div className="form_header">
          <h1 className="form_header__title">Basic info</h1>
          <div className="form_header__buttons">
            <button type="button" className="form_header__buttons___button button_white">
              Cancel
            </button>
            <button type="submit" className="form_header__buttons___button button_blue">
              {loading ? 'Updating...' : 'Update'}
            </button>
          </div>
        </div>
        <div className="form_content">
          <div className="form_content__inputField">
            <label htmlFor="first_name">
              <span className="form_content__inputField___label">FIRST NAME</span>
              <input
                type="text"
                className="form_content__inputField___input"
                name="first_name"
                defaultValue={firstName}
                placeholder="Enter your first name"
                onChange={this.handleChange}
              />
            </label>
          </div>
          <div className="form_content__inputField">
            <label htmlFor="last_name">
              <span className="form_content__inputField___label">LAST NAME</span>
              <input
                type="text"
                className="form_content__inputField___input"
                name="last_name"
                defaultValue={lastName}
                placeholder="Enter your last name"
                onChange={this.handleChange}
              />
            </label>
          </div>
          <div className="form_content__inputField">
            <label htmlFor="user_name">
              <span className="form_content__inputField___label">USERNAME</span>
              <input
                type="text"
                className="form_content__inputField___input"
                name="user_name"
                defaultValue={userName}
                placeholder="Enter your user name"
                onChange={this.handleChange}
              />
            </label>
          </div>
          <div className="form_content__inputField">
            <label htmlFor="gender">
              <span className="form_content__inputField___label">GENDER</span>
              <div className="form_content__inputField__select">
                <select
                  className="form_content__inputField___input"
                  name="gender"
                  defaultValue={gender}
                  onChange={this.handleChange}
                >
                  <option disabled selected className="placeholder">
                    Select your gender
                  </option>
                  <option value="Male">Male</option>
                  <option value="Female">Female</option>
                </select>
              </div>
            </label>
          </div>
          <div className="form_content__inputField">
            <label htmlFor="dob">
              <span className="form_content__inputField___label">DATE OF BIRTH</span>
              <input
                type="date"
                className="form_content__inputField___input"
                name="dob"
                defaultValue={moment(dob).format('YYYY-MM-DD')}
                onChange={this.handleChange}
              />
            </label>
          </div>
          <div className="form_content__inputField">
            <label htmlFor="phone">
              <span className="form_content__inputField___label">PHONE</span>
              <input
                type="text"
                className="form_content__inputField___input"
                name="phone"
                defaultValue={phone}
                placeholder="Enter your Phone number"
                onChange={this.handleChange}
              />
            </label>
          </div>
          <div className="form_content__inputField">
            <label htmlFor="address">
              <span className="form_content__inputField___label">ADDRESS</span>
              <input
                type="text"
                className="form_content__inputField___input"
                name="address"
                defaultValue={address}
                placeholder="Enter your current address"
                onChange={this.handleChange}
              />
            </label>
          </div>
          <div className="form_content__inputField">
            <label htmlFor="country">
              <span className="form_content__inputField___label">COUNTRY</span>
              <input
                type="text"
                className="form_content__inputField___input"
                name="country"
                defaultValue={country}
                placeholder="Enter your country"
                onChange={this.handleChange}
              />
            </label>
          </div>
          <div className="form_content__inputField">
            <label htmlFor="prefered_language">
              <span className="form_content__inputField___label">PREFERED LANGUAGE</span>
              <input
                type="text"
                className="form_content__inputField___input"
                name="prefered_language"
                defaultValue={preferedLanguage}
                placeholder="Enter your prefered language"
                onChange={this.handleChange}
              />
            </label>
          </div>
          <div className="form_content__inputField">
            <label htmlFor="prefered_currency">
              <span className="form_content__inputField___label">PREFERED CURRENCY</span>
              <input
                type="text"
                className="form_content__inputField___input"
                name="prefered_currency"
                defaultValue={preferdCurrency}
                placeholder="Enter your prefered currency"
                onChange={this.handleChange}
              />
            </label>
          </div>
          <div className="form_content__inputField">
            <label htmlFor="company">
              <span className="form_content__inputField___label">COMPANY</span>
              <input
                type="text"
                className="form_content__inputField___input"
                name="company"
                defaultValue={company}
                placeholder="Enter your company"
                onChange={this.handleChange}
              />
            </label>
          </div>
          <div className="form_content__inputField">
            <label htmlFor="department">
              <span className="form_content__inputField___label">DEPARTMENT</span>
              <input
                type="text"
                className="form_content__inputField___input"
                name="department"
                defaultValue={department}
                placeholder="Enter your department"
                onChange={this.handleChange}
              />
            </label>
          </div>
        </div>
      </form>
    );
  }
}

ProfileForm.propTypes = {
  updateProfile: PropTypes.func.isRequired,
  loading: PropTypes.bool,
  profile: PropTypes.objectOf(PropTypes.any),
};

ProfileForm.defaultProps = {
  loading: false,
  profile: {},
};

export default ProfileForm;
