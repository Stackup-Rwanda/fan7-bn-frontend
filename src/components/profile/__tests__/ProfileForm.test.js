/* eslint-disable react/jsx-props-no-spreading */
import React from 'react';
import { mount, shallow } from 'enzyme';
import ProfileForm from '../ProfileForm';

describe('ProfileForm Component Test', () => {
  it('should return true if component exists', () => {
    const wrapper = mount(<ProfileForm />);

    expect(wrapper.find('form').exists()).toBe(true);
  });

  it('should check if it is loading', () => {
    const props = {
      loading: true,
    };
    const wrapper = mount(<ProfileForm {...props} />);

    expect(wrapper.find('button[type="submit"]').text()).toEqual('Updating...');
  });

  it('should handle change', () => {
    const event = {
      persist: jest.fn(),
      target: {
        name: 'first_name',
        value: 'Elvis',
      },
    };
    const props = {
      loading: false,
    };
    const wrapper = shallow(<ProfileForm {...props} />);

    wrapper.find('input[name="first_name"]').simulate('change', event);

    expect(wrapper.state().user.first_name).toEqual('Elvis');
  });

  it('should handle submit', () => {
    const updateProfile = jest.fn();
    const event = {
      preventDefault: jest.fn(),
    };
    const props = {
      updateProfile,
    };
    const wrapper = shallow(<ProfileForm {...props} />);

    wrapper.find('form').simulate('submit', event);

    expect(updateProfile).toHaveBeenCalled();
  });
});

// describe('<ProfileForm /> shallow rendering tests', () => {
//   it('matches the snapshot', () => {
//     const props = {
//       updateProfile: jest.fn(),
//     };
//     const tree = shallow(<ProfileForm {...props} />);
//     expect(tree).toMatchSnapshot();
//   });
// });
