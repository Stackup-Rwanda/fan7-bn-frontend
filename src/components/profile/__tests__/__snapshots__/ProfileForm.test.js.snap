// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProfileForm /> shallow rendering tests matches the snapshot 1`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <div
    className="form_header"
  >
    <h1
      className="form_header__title"
    >
      Basic info
    </h1>
    <div
      className="form_header__buttons"
    >
      <button
        className="form_header__buttons___button button_white"
        type="button"
      >
        Cancel
      </button>
      <button
        className="form_header__buttons___button button_blue"
        type="submit"
      >
        Update
      </button>
    </div>
  </div>
  <div
    className="form_content"
  >
    <div
      className="form_content__inputField"
    >
      <label
        htmlFor="first_name"
      >
        <span
          className="form_content__inputField___label"
        >
          FIRST NAME
        </span>
        <input
          className="form_content__inputField___input"
          name="first_name"
          onChange={[Function]}
          placeholder="Enter your first name"
          type="text"
        />
      </label>
    </div>
    <div
      className="form_content__inputField"
    >
      <label
        htmlFor="last_name"
      >
        <span
          className="form_content__inputField___label"
        >
          LAST NAME
        </span>
        <input
          className="form_content__inputField___input"
          name="last_name"
          onChange={[Function]}
          placeholder="Enter your last name"
          type="text"
        />
      </label>
    </div>
    <div
      className="form_content__inputField"
    >
      <label
        htmlFor="user_name"
      >
        <span
          className="form_content__inputField___label"
        >
          USERNAME
        </span>
        <input
          className="form_content__inputField___input"
          name="user_name"
          onChange={[Function]}
          placeholder="Enter your user name"
          type="text"
        />
      </label>
    </div>
    <div
      className="form_content__inputField"
    >
      <label
        htmlFor="gender"
      >
        <span
          className="form_content__inputField___label"
        >
          GENDER
        </span>
        <div
          className="form_content__inputField__select"
        >
          <select
            className="form_content__inputField___input"
            name="gender"
            onChange={[Function]}
          >
            <option
              className="placeholder"
              disabled={true}
              selected={true}
            >
              Select your gender
            </option>
            <option
              value="Male"
            >
              Male
            </option>
            <option
              value="Female"
            >
              Female
            </option>
          </select>
        </div>
      </label>
    </div>
    <div
      className="form_content__inputField"
    >
      <label
        htmlFor="dob"
      >
        <span
          className="form_content__inputField___label"
        >
          DATE OF BIRTH
        </span>
        <input
          className="form_content__inputField___input"
          defaultValue="2020-03-25"
          name="dob"
          onChange={[Function]}
          type="date"
        />
      </label>
    </div>
    <div
      className="form_content__inputField"
    >
      <label
        htmlFor="phone"
      >
        <span
          className="form_content__inputField___label"
        >
          PHONE
        </span>
        <input
          className="form_content__inputField___input"
          name="phone"
          onChange={[Function]}
          placeholder="Enter your Phone number"
          type="text"
        />
      </label>
    </div>
    <div
      className="form_content__inputField"
    >
      <label
        htmlFor="address"
      >
        <span
          className="form_content__inputField___label"
        >
          ADDRESS
        </span>
        <input
          className="form_content__inputField___input"
          name="address"
          onChange={[Function]}
          placeholder="Enter your current address"
          type="text"
        />
      </label>
    </div>
    <div
      className="form_content__inputField"
    >
      <label
        htmlFor="country"
      >
        <span
          className="form_content__inputField___label"
        >
          COUNTRY
        </span>
        <input
          className="form_content__inputField___input"
          name="country"
          onChange={[Function]}
          placeholder="Enter your country"
          type="text"
        />
      </label>
    </div>
    <div
      className="form_content__inputField"
    >
      <label
        htmlFor="prefered_language"
      >
        <span
          className="form_content__inputField___label"
        >
          PREFERED LANGUAGE
        </span>
        <input
          className="form_content__inputField___input"
          name="prefered_language"
          onChange={[Function]}
          placeholder="Enter your prefered language"
          type="text"
        />
      </label>
    </div>
    <div
      className="form_content__inputField"
    >
      <label
        htmlFor="prefered_currency"
      >
        <span
          className="form_content__inputField___label"
        >
          PREFERED CURRENCY
        </span>
        <input
          className="form_content__inputField___input"
          name="prefered_currency"
          onChange={[Function]}
          placeholder="Enter your prefered currency"
          type="text"
        />
      </label>
    </div>
    <div
      className="form_content__inputField"
    >
      <label
        htmlFor="company"
      >
        <span
          className="form_content__inputField___label"
        >
          COMPANY
        </span>
        <input
          className="form_content__inputField___input"
          name="company"
          onChange={[Function]}
          placeholder="Enter your company"
          type="text"
        />
      </label>
    </div>
    <div
      className="form_content__inputField"
    >
      <label
        htmlFor="department"
      >
        <span
          className="form_content__inputField___label"
        >
          DEPARTMENT
        </span>
        <input
          className="form_content__inputField___input"
          name="department"
          onChange={[Function]}
          placeholder="Enter your department"
          type="text"
        />
      </label>
    </div>
  </div>
</form>
`;
